src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -0,0 +1,423 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Text;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal class DecisionDagBuilder
+    {
+        private readonly Conversions Conversions;
+        private readonly TypeSymbol BooleanType;
+        private readonly TypeSymbol ObjectType;
+
+        internal DecisionDagBuilder(CSharpCompilation compilation)
+        {
+            this.Conversions = compilation.Conversions;
+            this.BooleanType = compilation.GetSpecialType(SpecialType.System_Boolean);
+            this.ObjectType = compilation.GetSpecialType(SpecialType.System_Object);
+        }
+
+        public BoundDecisionDag CreateDecisionDag(BoundPatternSwitchStatement statement)
+        {
+            ImmutableArray<PartialCaseDecision> cases = MakeCases(statement);
+            BoundDecisionDag dag = MakeDecisionDag(cases);
+            return dag;
+        }
+
+        public BoundDagTemp LowerPattern(
+            BoundExpression input,
+            BoundPattern pattern,
+            out ImmutableArray<BoundDagDecision> decisions,
+            out ImmutableArray<(BoundExpression, BoundDagTemp)> bindings)
+        {
+            var rootIdentifier = new BoundDagTemp(input.Syntax, input.Type, null, 0);
+            MakeAndSimplifyDecisionsAndBindings(rootIdentifier, pattern, out decisions, out bindings);
+            return rootIdentifier;
+        }
+
+        private ImmutableArray<PartialCaseDecision> MakeCases(BoundPatternSwitchStatement statement)
+        {
+            var rootIdentifier = new BoundDagTemp(statement.Expression.Syntax, statement.Expression.Type, null, 0);
+            int i = 0;
+            var builder = ArrayBuilder<PartialCaseDecision>.GetInstance();
```

Comment:
>GetInstance [](start = 60, length = 11)

Consider if we can guess the size for the common case.

Reply:
Addressed (by using a minimum size) in [https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec](https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec) of [https://github.com/dotnet/roslyn/pull/23357](https://github.com/dotnet/roslyn/pull/23357)

---
In reply to: [156532115](https://github.com/dotnet/roslyn/pull/23209#discussion_r156532115) [](ancestors = 156532115)